src/nvapi_gpu.cpp

```diff
@@ -382,11 +382,106 @@ extern "C" {
         if (!nvapiAdapterRegistry->IsAdapter(adapter))
             return ExpectedPhysicalGpuHandle(n);
 
-        *pSize = adapter->GetVirtualVRamSize();
+        *pSize = (adapter->GetVideoMemory().DedicatedVideoMemory + adapter->GetVideoMemory().DedicatedSystemMemory) / 1024;
 
         return Ok(n);
     }
 
+    NvAPI_Status __cdecl NvAPI_GPU_GetMemoryInfo(NvPhysicalGpuHandle hPhysicalGpu, NV_DISPLAY_DRIVER_MEMORY_INFO* pMemoryInfo) {
+        constexpr auto n = __func__;
+        thread_local bool alreadyLoggedOk = false;
+
+        if (log::tracing())
+            log::trace(n, log::fmt::hnd(hPhysicalGpu), log::fmt::ptr(pMemoryInfo));
+
+        if (nvapiAdapterRegistry == nullptr)
+            return ApiNotInitialized(n);
+
+        if (pMemoryInfo == nullptr)
+            return InvalidArgument(n);
+
+        auto adapter = reinterpret_cast<NvapiAdapter*>(hPhysicalGpu);
+        if (!nvapiAdapterRegistry->IsAdapter(adapter))
+            return ExpectedPhysicalGpuHandle(n);
+
+        adapter->UpdateVideoMemory();
+
+        switch (pMemoryInfo->version) {
+            case NV_DISPLAY_DRIVER_MEMORY_INFO_VER_1: {
+                auto pMemoryInfoV1 = reinterpret_cast<NV_DISPLAY_DRIVER_MEMORY_INFO_V1*>(pMemoryInfo);
+                pMemoryInfoV1->dedicatedVideoMemory = adapter->GetVideoMemory().DedicatedVideoMemory;
+                pMemoryInfoV1->availableDedicatedVideoMemory = adapter->GetVideoMemory().Budget;
+                pMemoryInfoV1->systemVideoMemory = adapter->GetVideoMemory().DedicatedSystemMemory;
+                pMemoryInfoV1->sharedSystemMemory = adapter->GetVideoMemory().SharedSystemMemory;
+                break;
+            }
+            case NV_DISPLAY_DRIVER_MEMORY_INFO_VER_2: {
+                auto pMemoryInfoV2 = reinterpret_cast<NV_DISPLAY_DRIVER_MEMORY_INFO_V2*>(pMemoryInfo);
+                pMemoryInfoV2->dedicatedVideoMemory = adapter->GetVideoMemory().DedicatedVideoMemory;
+                pMemoryInfoV2->availableDedicatedVideoMemory = adapter->GetVideoMemory().Budget;
+                pMemoryInfoV2->systemVideoMemory = adapter->GetVideoMemory().DedicatedSystemMemory;
+                pMemoryInfoV2->sharedSystemMemory = adapter->GetVideoMemory().SharedSystemMemory;
+                pMemoryInfoV2->curAvailableDedicatedVideoMemory = adapter->GetVideoMemory().Budget - adapter->GetVideoMemory().CurrentUsage;
```

Comment:
Not sure about this one. The others seem to roughly match.

Reply:
Do I understand correctly that:

* `dedicatedVideoMemory` is simply the total size of device-local heap and should probably never change,
* `availableDedicatedVideoMemory` is how much of above is left for the application to allocate, so `dedicatedVideoMemory` reduced by other application's allocations and areas reserved for private use by the driver, if any (and current application allocating memory shouldn't affect this),
* `curAvailableDedicatedVideoMemory` is how much of above is left for the application to allocate after we include allocations it already made, so `availableDedicatedVideoMemory` reduced by application's own allocated memory size?

Reply:
This is indeed more or less how I understood the terminology, based on the nvapi headers and @SveSop 's numbers:

```c++
    NvU64   dedicatedVideoMemory;              //!< Size(in bytes) of the physical framebuffer. Refers to the dedicated video memory on discrete GPUs.
                                               //!  It is more performant for GPU operations than the reserved systemVideoMemory.
    NvU64   availableDedicatedVideoMemory;     //!< Size(in bytes) of the available physical framebuffer for allocating video memory surfaces.
    NvU64   systemVideoMemory;                 //!< Size(in bytes) of system memory the driver allocates at load time. It is a substitute for dedicated video memory.
                                               //!< Typically used with integrated GPUs that do not have dedicated video memory.
    NvU64   sharedSystemMemory;                //!< Size(in bytes) of shared system memory that driver is allowed to commit for surfaces across all allocations.
                                               //!< On discrete GPUs, it is used to utilize system memory for various operations. It does not need to be reserved during boot.
                                               //!< It may be used by both GPU and CPU, and has an �on-demand� type of usage.
    NvU64   curAvailableDedicatedVideoMemory;  //!< Size(in bytes) of the current available physical framebuffer for allocating video memory surfaces. 
```